### ISSUE TITLE
Authentication with LDAP user

### ISSUE THREAD

**Issue by kichetof** [litstack/litstack#103]
Hi guys,

I discovered your package last week and really love it! It really simple to use it, awesome! Many Thanks!!

One of the feature that I would like to implement, is the relation between my app and my AD. For this I use [LdapRecord for Laravel](https://github.com/DirectoryTree/LdapRecord-Laravel) that use [LdapRecord](https://ldaprecord.com/). It works really well with a fresh install of Laravel with Jetstream that used Fortify. LdapRecord is [implemented](https://ldaprecord.com/docs/laravel/v1/auth/laravel-jetstream/) through it.

Do you see a possibility to implement it with Litstack ? To do it more complex, I need to use [username](https://ldaprecord.com/docs/laravel/v1/auth/laravel-jetstream/#fortify-setup-username-config) instead of mail who is sAMAccountName in AD.

I hope that you could help me to implement it :)

Many thanks for your great package and your help!!
Enjoy and take care!
Christophe

(PS: I'm like a child before Xmas to see [that](https://github.com/litstack/litstack/issues/80#issuecomment-713491800) in action)

**Comment by cbl** [comment]
Hey @kichetof, happy to hear you like litstack!

In https://github.com/litstack/litstack/pull/90 I added custom authorization to litstack. However I haven't documented this yet on [litstack.io](https://litstack.io).
I see that it is not possible at the moment to switch between username and email. I'll propably add this in the future however I would be happy a about a pr as well.

**Comment by kichetof** [comment]
Hi @cbl! Thanks for your quick reply!

I'll try to look where I can add the Auth::attempt and try with username instead of samaccountname for the moment (It should works if LdapRecord is correctly configured). Need to learn more Laravel auth :)

**Comment by kichetof** [comment]
After some tests, I could have a semi working solution.

```
// Ignite\Auth\Authentication.php

// function attempt( ...
// ...
        $credentials['username'] = $credentials['email'];
        $credentials['samaccountname'] = $credentials['email'];
        unset($credentials['email']);
        unset($credentials['username']);
// ...
```
TODO *(tried but doesn't work)*: 
set a config var (exemple) `lit.auth.username = 'samaccountname'` and use it on Authentication.php and login.blade.php but it doesn't work, I don't know why :thinking:

In my User model I need to add: `use Notifiable, HasRoles, CanResetPassword;`

And when I try to login, connection with AD is successful :star_struck: *(tried with a user that was not already in my db user)* and I get this error right after:
```
ErrorException
Argument 1 passed to Lit\Http\Controllers\User\UserController::authorize() must be an instance of Lit\Models\User, instance of App\Models\User given, called in /var/www/test/litstack/vendor/litstack/litstack/src/Crud/Config/Traits/HasCrudShow.php on line 54 (View: /var/www/test/litstack/vendor/litstack/litstack/resources/views/index.blade.php) (View: /var/www/test/litstack/vendor/litstack/litstack/resources/views/index.blade.php)
```

I think that the solution is not to far away!!

**Comment by cbl** [comment]
Hey @kichetof,

switch the ` Lit\Models\User` in the authorize method with the `Illuminate\Contracts\Auth\Access\Authorizable` contract. This should fix it. You may also need to switch it in the `Lit\Http\Controllers\User\ProfileSettingsConfig` controller.

Also if you want to use the build in profile settings you should change the `model` property in `Lit\Config\User\ProfileSettingsConfig` and `Lit\Config\User\UserConfig` to `App\Models\User`.

**Comment by cbl** [comment]
Also the only thing your user model need's is the `Illuminate\Contracts\Auth\Access\Authorizable` and the corresponding trait `Illuminate\Foundation\Auth\Access\Authorizable`.

**Comment by kichetof** [comment]
Thanks for your help!
I added `use Illuminate\Contracts\Auth\Access\Authorizable;` in my User model and changed `Lit\Models\User` in theses file bypass the last error:

- `Lit\Http\Controllers\User\UserController`
- `Lit\Config\User\ProfileSettingsConfig`
- `Lit\Config\User\UserConfig`

Next I'll try to catch this error: `Call to a member function icon() on null (View: /var/www/test/litstack/vendor/litstack/litstack/resources/views/index.blade.php) (View: /var/www/test/litstack/vendor/litstack/litstack/resources/views/index.blade.php)` and try to add a config for the username.

**Comment by cbl** [comment]
Hey, seem's like we have a bug here.

You could comment out the following line in your `NavigationConfig`:

```php
$nav->preset('user.user')->icon(fa('users')),
```

Or change it to:
```php
$nav->preset('user.user', ['icon' => fa('users')]),
```

**Comment by kichetof** [comment]
Yes I changed my `NavigationConfig` to 
```php
$nav->preset('user.user', ['icon' => fa('users')]),
```
And the error go away!

I tried to start from scratch with a new instance of Laravel + Litstack + LdapRecord to setup from 0. Login works well but I'm still with my white screen with Litstack logo on the left + round loading. Need more investigations to finally works :)

**Comment by cbl** [comment]
Should be fixed. Release tagged in [v3.2.3](https://github.com/litstack/litstack/releases/tag/v3.2.3).

**Comment by kichetof** [comment]
Hi guys!
Thanks for fixing this issue!
Now to be back with the main question, we need to custom user field to identify with the guard.
Something like that:

```php
// src/Auth/Authentication.php

public function attempt($credentials, bool $remember = false, array $parameters = [])
    {
        if ($this->guard->attempt($credentials, $remember)) {
            return $this->extendedAttempt($parameters);
        }

        if (! config('lit.login.username')) {
            return false;
        }

        // use a config value to store the field name
        $credentials[config('lit.login.username_field')] = $credentials['email']; <-- this could be used everywhere instead of email
        unset($credentials['email']);

        if ($this->guard->attempt($credentials, $remember)) {
            return $this->extendedAttempt($parameters);
        }

        return false;
    }


// config/lit.php

'login' => [
        'username' => true,
        'username_field' => 'samaccountname',
    ],
```

---

Another issue that I just discover, I've no admin permission. I tried to edit from mysql to add role but didn't work.

role_id | model_type | model_id
------ | ------------ | ---------
1 | App\Models\User | 1

But I can't access to roles / permission from the menu, do you've any ideas?

**Comment by lpheller** [comment]
@kichetof Regarding the permissions: Try renaming the `guard_name` for all `roles` and `permissions` entries to fit the guard name in your lit config.

**Comment by kichetof** [comment]
@lpheller `roles` and `permissions` has been changed, table `role_has_permissions` was filled with rights ids and I can't access to permissions manager.
Do you have some ideas?

**Comment by lpheller** [comment]
I made a quick example project setup thats using the default `web` guard and `App\Models\User`. Maybe you want to take a look at it: https://github.com/lpheller/litstack-authguard-test

This way its also working if you delete the entry within `model_has_roles` and add your entries to `model_has_permissions` like this
```sql
INSERT INTO `model_has_permissions` (`permission_id`, `model_type`, `model_id`)
VALUES (1, 'App\\Models\\User', 1);
```

**Comment by kichetof** [comment]
Yeah that works fine! 8)

Unfortunately, info from Users are not displayed.
<img width="1055" alt="Capture d’écran 2020-12-18 à 10 15 23" src="https://user-images.githubusercontent.com/12150996/102596828-0b54e400-411a-11eb-993b-909df0cd08ae.png">

Would it be possible to disable "Create user" as user are binding from our AD ?

**Comment by cbl** [comment]
You can edit the UserConfig index table columns `lit/app/Config/User/UserConfig` https://github.com/litstack/litstack/blob/03b2a73422e4be24df2788ef4c509317a6e57a02/publish/lit/app/Config/User/UserConfig.php#L75

**Comment by lpheller** [comment]
That's probably just because the default laravel user doesn't provide the same attributes as the lit_user model. You might want to add these attributes or just update the `lit/app/Config/User/UserConfig` to fit your model.

**Comment by kichetof** [comment]
Yeah you rock guys! That was the key! My user table isn't the same (username and name)!

We only need a way to customize the field to login and my issue will be successfully closed :D
